### ISSUE TITLE
Tasks stuck in queued state

### ISSUE THREAD

**Issue by Chris7** [apache/airflow#21225]
### Apache Airflow version

2.2.3 (latest released)

### What happened

Tasks are stuck in the queued state and will not be scheduled for execution. In the logs, these tasks have a message of `could not queue task <task details>`, as they are currently in the `queued` or `running` sets in the executor.

### What you expected to happen

Tasks run :)

### How to reproduce

We have a dockerized airflow setup, using celery with a rabbit broker & postgres as the result db. When rolling out DAG updates, we redeploy most of the components (the workers, scheduler, webserver, and rabbit). We can have a few thousand Dagruns at a given time. This error seems to happen during a load spike when a deployment happens.

Looking at the code, this is what I believe is happening:

Starting from the initial debug message of `could not queue task` I found tasks are marked as running (but in the UI they still appear as queued): https://github.com/apache/airflow/blob/main/airflow/executors/base_executor.py#L85

Tracking through our logs, I see these tasks are recovered by the adoption code, and the state there is STARTED (https://github.com/apache/airflow/blob/main/airflow/executors/celery_executor.py#L540). 

Following the state update code, I see this does not cause any state updates to occur in Airflow (https://github.com/apache/airflow/blob/main/airflow/executors/celery_executor.py#L465). Thus, if a task is marked as STARTED in the results db, but queued in the airflow task state, it will never be transferred out by the scheduler. However ,you can get these tasks to finally run by clicking the `run` button. 

### Operating System

Ubuntu 20.04.3 LTS

### Deployment

Other Docker-based deployment

### Anything else

I believe this could be addressed by asking celery if a task is actually running in `try_adopt_task_instances`. There are obvious problems like timeouts, but `celery inspect active|reserved` can return a json output of running and reserved tasks to verify a STARTED task is actually running

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by dhananjays** [comment]
We're facing this as well after upgrading to `2.2.3` from `2.1.0`

**Comment by dimon222** [comment]
Plus one here, noticing this frequently after bump from 2.1.4 to 2.2.3

**Comment by Chris7** [comment]
This is not ideal, but for those who want to kick all their queued tasks out, here's a snippet i've been using:
```
from airflow import models, settings
session = settings.Session()

from airflow.executors.executor_loader import ExecutorLoader
tis = session.query(models.TaskInstance).filter(models.TaskInstance.state=='queued')
dagbag = models.DagBag()
for ti in tis:
  dag = dagbag.get_dag(ti.dag_id)
  task = dag.get_task(ti.task_id)
  ti.refresh_from_task(task)
  executor = ExecutorLoader.get_default_executor()
  executor.job_id = "manual"
  executor.start()
  executor.queue_task_instance(ti, ignore_all_deps=False, ignore_task_deps=False, ignore_ti_state=False)
  executor.heartbeat()
```

**Comment by potiuk** [comment]
I think this is being handled in one of the fixes coming to 2.2.4 (@ephraimbuddy ?)

**Comment by ephraimbuddy** [comment]
> I think this is being handled in one of the fixes coming to 2.2.4 (@ephraimbuddy ?)

Yes. Here https://github.com/apache/airflow/pull/19769

**Comment by meetri** [comment]
i'm still having this issue on 2.2.5 any updates on this ticket or when it may get fixed. In the mean time i'm using the code snippet provided by @Chris7

**Comment by potiuk** [comment]
> i'm still having this issue on 2.2.5 any updates on this ticket or when it may get fixed. In the mean time i'm using the code snippet provided by @Chris7

This issue is marked as fixed. You likely experience other issue. But we will never know unless you open a new issue and provide all the details, your logs and circumstances that you experience the problem. Unless you do that - there is no chance your issue will be handled because no-one has any idea about Your issue.

**Comment by potiuk** [comment]
Just be aware @meetri that you get the software for free - and a lot of the people contribute to it for free. And the best value you can bring is to spend some time on detailing your problems. providing as much as you can - including the circumstances, details and reproducibility.  This way the smallest help you can provide for those who try to help people like you - on a free forum, for - again - software that you paid precisely 0 USD for.

**Comment by vandanthaker** [comment]
https://github.com/apache/airflow/issues/21225#issuecomment-1035656930
@arkadiusz-bach - how did you fix it? I tried changing queue name from default kubernetes to kubernetesQ like below
```
[celery_kubernetes_executor]
kubernetes_queue = kubernetesQ
```

But I still see K8S executor picking up tasks from default queue(celery) which are in queued state
```
[[34m2022-04-18 11:10:05,800[0m] {[34mkubernetes_executor.py:[0m454} INFO[0m - Found 2 queued task instances[0m
[[34m2022-04-18 11:10:05,823[0m] {[34mkubernetes_executor.py:[0m491} INFO[0m - TaskInstance: <TaskInstance: dbr_concurrent_trigger_poll.gcorp_ccl.trigger_descision_for_source scheduled__2022-04-18T11:05:00+00:00 [queued]> found in queued state but was not launched, rescheduling[0m
```

Thanks in advance

**Comment by arkadiusz-bach** [comment]
In kubernetes_executor.py:

- At the beginning import conf and SQLAlchemy 'and_' operator:
  ```
  from airflow.configuration import conf
  from sqlalchemy import and_
  ```
- Inside \_\_init\_\_ of KubernetesExecutor class:
   ```
    self.is_celery_kubernetes_executor = conf.get('core', 'executor') == 'CeleryKubernetesExecutor': 
    self.kubernetes_queue = conf.get('celery_kubernetes_executor', 'kubernetes_queue')
   ```
 - Replace
    https://github.com/apache/airflow/blob/5a6a2d604979cb70c5c9d3797738f0876dd38c3b/airflow/executors/kubernetes_executor.py#L455
   With:
   ```
    if self.is_celery_kubernetes_executor:
        queued_tasks = session.query(TaskInstance).filter(and_(TaskInstance.state = State.QUEUED, TaskInstance.queue == self.kubernetes_queue)).all()
    else:
        queued_tasks = session.query(TaskInstance).filter(TaskInstance.state = State.QUEUED).all()
   ```

PS: I just wrote that in notepad, so please check whether there are no syntax errors

**Comment by vandanthaker** [comment]
Got it! see similar fix in PR too. so no workaround w/o code patching, which is also evident as the query(inside k8s executor) on task instances is looking only queued state and not queue(celery/k8s). Thanks

**Comment by meetri** [comment]
> Just be aware @meetri that you get the software for free - and a lot of the people contribute to it for free. And the best value you can bring is to spend some time on detailing your problems. providing as much as you can - including the circumstances, details and reproducibility. This way the smallest help you can provide for those who try to help people like you - on a free forum, for - again - software that you paid precisely 0 USD for.

Understood, and thanks you to everyone that contributes to this project. If needed I can provide whatever information to track down this issue. I posted my comment only because I thought it would be of use for anyone passing through that even the latest version of the code has this issue and how I got around it for the time being.

This may or may not be of help but I did since my post tune things to such an extent to where I didn't need to run the script posted by Chris7, https://github.com/apache/airflow/issues/21225#issuecomment-1029060830

Firstly, I'm running this in amazon in an EKS cluster.

basically, the biggest issue was I didn't have enough memory allocated to the scheduler by a lot. I bumped it up to about 8GB minimum. This stopped a lot of the errors related to the scheduler missing heartbeats.

Secondly, i upped the number of cores at minimum to 2 per container. I am running 2 airflow scheduler replicas.

Lastly, I'm not sure how big of an impact this was probably pretty major. But given my pods autoscale up and down on demand based on a redis queue. I believe that pods were getting scaled down and the workers were being force terminated. Once i added a dumb-init to the airflow worker docker image that problem seemed to go away.

I'm only posting this if anyone is having similar issues. This may or may not be a bug in the airflow.

**Comment by nuclear0wl** [comment]
hello @meetri, could you please provide more info about you modification to Docker image? I also mentioned that I'm getting "lost" Queued tasks when turning on Redis-based autoscaling but can't understand how some kind of *dumb-init* can help to avoid an issue in this situation.

**Comment by eladkal** [comment]
Is anyone experienced this issue after 2.3.1?
I would imagine that https://github.com/apache/airflow/pull/23690 should make things better... even if not fixing the issue but making sure that tasks don't get stuck in queue state.

**Comment by patricker** [comment]
We have been sing tasks stuck in a queued state on Airflow 2.3.2. Not many of them, but a few just get stuck in Queued and I have to go either clear them or run them.

**Comment by potiuk** [comment]
@patricker - can you please update to 2.3.4 and see it solves your problem? You will have to do it sooner or later (and it has dozens of fixes) and might be fastest way for you (and us) to fix your problem.

**Comment by buffaloDeveloper** [comment]
Seeing the same issue using bitnami's airflow image: docker.io/bitnami/airflow-scheduler:2-debian-11, which I believe is mapped to 2.3.4.  I had to downgrade to 2.3.3-debian-11-r1 and it's working fine again.

**Comment by V0lantis** [comment]
Hello there 👋 

We experienced the same issue in our deployment cluster, and after a lot of searching, I think I have found why it is happening as some reproductible steps to provoke it. I am very willing to submit a PR if I may.

First off, we experienced this issue very rarely on our deployment cluster. 
Our deployment cluster is right now on **AWS EKS**, Airflow version **2.3.3**, with **redis** message broker and **CeleryExecutor** of course :). 
After a lot of searching, I found out that the queued task (or the queued tasks, I don't know how many we had because we have 1000+ dags and even more tasks), which didn't want to get executed appeared when our AWS autoscaler was doing a downscaling of our pod. At that point, when I looked at the two pods celery workers which were downscaled, I saw a a log message in the pod : 

```
worker: Warm shutdown
```
this log message appeared a couple of milliseconds after the scheduler has sent the task to redis.
What if, the worker had consumed this message, but in the meantime, celery shutdown the consumer loop, and the message never get properly executed. **BUT**, since the message has been consumed, (LPOP), redis didn't have the message in its queue, but the Celery Executor still had it ! That's why we have the log (after recleaning the task):

```
[2022-08-27 21:15:40,100] {base_executor.py:93} INFO - Adding to queue: ['airflow', 'tasks', 'run', 'example_bash_operator', 'also_run_this', 'manual__2022-08-27T19:13:04.614822+00:00', '--local', '--subdir', '/Users/arthurvolant/dev/python/airflow/airflow/example_dags/example_bash_operator.py']
[2022-08-27 21:15:40,101] {base_executor.py:213} INFO - task TaskInstanceKey(dag_id='example_bash_operator', task_id='also_run_this', run_id='manual__2022-08-27T19:13:04.614822+00:00', try_number=1, map_index=-1) is still running
[2022-08-27 21:15:40,205] {base_executor.py:213} INFO - task TaskInstanceKey(dag_id='example_bash_operator', task_id='also_run_this', run_id='manual__2022-08-27T19:13:04.614822+00:00', try_number=1, map_index=-1) is still running
[2022-08-27 21:15:41,292] {base_executor.py:213} INFO - task TaskInstanceKey(dag_id='example_bash_operator', task_id='also_run_this', run_id='manual__2022-08-27T19:13:04.614822+00:00', try_number=1, map_index=-1) is still running
[2022-08-27 21:15:42,391] {base_executor.py:213} INFO - task TaskInstanceKey(dag_id='example_bash_operator', task_id='also_run_this', run_id='manual__2022-08-27T19:13:04.614822+00:00', try_number=1, map_index=-1) is still running
[2022-08-27 21:15:43,498] {base_executor.py:217} ERROR - could not queue task TaskInstanceKey(dag_id='example_bash_operator', task_id='also_run_this', run_id='manual__2022-08-27T19:13:04.614822+00:00', try_number=1, map_index=-1) (still running after 4 attempts)
```

I guess, that's why there is a `visibility_timeout`, but unfortunately, ours is 25hours and some of our tasks cannot be suffer any delay (maximum half an our). Anyway, here are the steps to reproduce it :

1. Start a basic cluster / docker dev env (it still works on the latest commit, Airflow **2.3.4.dev** `breeze start-airflow -p 3.10 -b postgres  --integration redis --db-reset`
2. Start a dag (In my example, I am starting _example_bash_operator_.
3. Tasks will be scheduled and put into queue by the CeleryExecutor
4. Connect to redis and LPOP one or more messages : `docker exec -it docker-compose_redis_1 redis-cli` and `LPOP default`
5. Activate the celeryWorker.
6. Boom 💥 . All the tasks which are still in queued are run, except the one(s) you popped out of the queue.

I will try to think of something. I have some ideas and submit a PR as soon as possible since this is very stressful for our deployment.

**Comment by christophedcpm** [comment]
I'm using airflow 2.2.5 in cloud-composer 2.0.24 on GCP and I have noticed the same kind of error. My investigation also lead me to think that the issue happens when the autoscaler stop worker pods.

**Comment by arkadiusz-bach** [comment]
@V0lantis task might stuck in the queued state at least due to:
- Redis crashed after receiving the task and it had no time / was not configured to save its state to disk, so the task was lost, but scheduler thinks that task is waiting to be picked up workers
- terminationGracePeriodSeconds  on your worker PODs is too low or it is not there at all.(default is 60 seconds)
 
This message `worker: Warm shutdown` means that celery received SIGTERM signal and it started gracefull shutdown - it is not going to pick any more tasks from redis queue and it will wait for all of the running tasks to finish

But if you've got some tasks that may be running for longer than terminationGracePeriod then Kubernetes might send SIGKILL first and:
- Celery will not be able to wait for all of the running tasks to finish(those will end with failed status)
- it may be able to pick the task from queue, but not able to change it state to running(maybe your case)

Also some of the celery workers might receive SIGKILL signal, when there is not enough memory allocated and it may led to the same behaviour, unfortunately you may not see OOM events in the kubernetes cluster when it happens, becaue when there is more than one process running on the container in Kubernetes then it is chosing randomly one of the child processes within container and sends SIGKILL(Celery is running with Main process and child processes(workers)).

If the Main process receives SIGKILL you will probably see OOM event, but if child then tasks it was processing will fail(in the logs you will be able to see that it received SIGKILL singal) or stuck in queued state if it was able to pick it

**Comment by V0lantis** [comment]
Hey @arkadiusz-bach,

I am sorry, that's true that I haven't posted any logs, maybe I haven't been clear as well... 

> @V0lantis task might stuck in the queued state at least due to:
> 
> * Redis crashed after receiving the task and it had no time / was not configured to save its state to disk, so the task was lost, but scheduler thinks that task is waiting to be picked up workers

We use ElasticCache service. The service is configured to store data on disk and is too sized for the actual amount of tasks it received and as stated above, celery actually acknowledged that task was successfully sent to queue. But why not? I haven't checked the log in redis, but I would think I would see an error in celery worker.

> * terminationGracePeriodSeconds  on your worker PODs is too low or it is not there at all.(default is 60 seconds)

The warm shutdown was received 10 minutes before the pod was actually killed. But the task was sent and acknowledged by `CeleryExecutor` with:
```
 Setting external_id for <TaskInstance: example_bash_operator.this_will_skip scheduled__2022-08-28T00:00:00+00:00 [queued]> to 979a100a-5012-4135-b01f-b93577b39ede
``` 
in the scheduler_log, which actually means that the task was received by redis (or celery would through a connection error).

> This message `worker: Warm shutdown` means that celery received SIGTERM signal and it started gracefull shutdown - it is not going to pick any more tasks from redis queue and it will wait for all of the running tasks to finish

Yep, but as I said in my comment : 

> this log message (`worker: Warm shutdown`) appeared a **couple of milliseconds AFTER** the scheduler has sent the task to redis.

⬆️ I still think that Celery consumer exited during the handling of my message [here](https://github.com/celery/kombu/blob/master/kombu/transport/redis.py#L356-L370) and didn't have time to put it in the `unacked` hash.

> But if you've got some tasks that may be running for longer than terminationGracePeriod then Kubernetes might send SIGKILL first and:
> 
> * Celery will not be able to wait for all of the running tasks to finish(those will end with failed status)
> * it was able to pick the task from queue(before SIGTERM), but not able to change its state to running(After SIGKILL) - maybe your case

⬆️ I stil thinks that my pods received SIGKILL 10 minutes after the SIGTERM but I haven't been yet able to check the `terminationGracePeriod`  

> Also some of the celery workers might receive SIGKILL signal, when there is not enough memory allocated and it may led to the same behaviour, unfortunately you may not see OOM events in the kubernetes cluster when it happens, becaue when there is more than one process running on the container in Kubernetes then it is chosing randomly one of the child processes within container and sends SIGKILL(Celery is running with Main process and child processes(workers)).

Already check as well. We have a Grafana dash to monitor the hardware resources that the pod is consuming, and the workers were far below the threshold we gave them

> If the Main process receives SIGKILL you will probably see OOM event(container will restart/stop), but if child then tasks it was processing will fail(in the logs you will be able to see that it received SIGKILL singal) or stuck in queued state if it was able to pick it and container will be running fine, - one of the celery processes will be restarted

The thing is, the task never started. Thanks anyway for your input, it was really enriching :)

**Comment by V0lantis** [comment]
I have been trying to fix this issue in [celery/kombu](https://github.com/celery/kombu) where I beleive(d) was lying the issue we are talking about here (namely, that a SIGTERM was sent while the redis queue was consumed, leading to a message never being acknowledged and a message lost in the wild). 

The issue is, I have been trying to replicate the issue in kombu without any success. I managed to reproduce the issue directly in Airflow by `RPOP` a msg before calling celery worker and it worked perfectly as described at the beginning of this issue. But, at least in its last version (Celery **v.5.2.7**) which is the one we have in production, I can't reproduce the behavior (Sending a SIGTERM just after kombu has `BRPOP` the msg). The msg is rightfully acked, and rightfully put back in the queue after SIGTERM had been catched by python. 

I don't really know where to go from here now. 

I am putting the logs msg we have been having in our prod environment. Maybe someone else will have a better idea : 

**Scheduler logs**
<details>

```
	<TaskInstance: dag_id.task_id scheduled__2022-08-24T15:00:00+00:00 [scheduled]>
	<TaskInstance: dag_id.task_id scheduled__2022-08-24T15:00:00+00:00 [scheduled]>
[2022-08-24 16:23:50,060] {cheduler_job.py:546} INFO - Sending TaskInstanceKey(dag_id='dag_id', task_id='task_id', run_id='scheduled__2022-08-24T15:00:00+00:00', try_number=1, map_index=-1) to executor with priority 5 and queue default
[2022-08-24 16:23:50,060] {base_executor.py:91} INFO - Adding to queue: ['airflow', 'tasks', 'run', 'dag_id', 'task_id', 'scheduled__2022-08-24T15:00:00+00:00', '--local', '--subdir', 'DAGS_FOLDER/dag_id.py']
[2022-08-24 16:23:50,264] {cheduler_job.py:605} INFO - Executor reports execution of dag_id.task_id run_id=scheduled__2022-08-24T15:00:00+00:00 exited with status queued for try_number 1
[2022-08-24 16:23:50,276] {cheduler_job.py:632} INFO - Setting external_id for <TaskInstance: dag_id.task_id scheduled__2022-08-24T15:00:00+00:00 [queued]> to 139960c0-263d-4891-b1b0-712b077a0d2b
	<TaskInstance: dag_id.task_id scheduled__2022-08-24T15:00:00+00:00 [scheduled]>
	<TaskInstance: dag_id.task_id scheduled__2022-08-24T15:00:00+00:00 [scheduled]>
```

</details>

There were two workers which were downscaled with similar logs: 

**1rst worker**

<details>

```
[2022-08-24 16:23:49,597: WARNING/ForkPoolWorker-254] unknown id type, added to user and store ids columns: create_account_step.fingerprint

worker: Warm shutdown (MainProcess)
[2022-08-24 16:23:49,777: WARNING/ForkPoolWorker-254] unknown id type, added to user and store ids columns: login.fingerprint
```

</details>

The log above is not relevant. The relevant information here is that the worker received shutdown SIGTERM signal around 0.3 seconds before the task id was actually sent.

**Comment by potiuk** [comment]
I think a really good way to start wil be to upgrade Airflow to 2.4.0 once it is out (even today possibly?) and install it with the right constraints (so that the right celery /kombu packages get installed). That sounds like it could have been handled already in the meantime by some of the recent Airflow fixes or - more likely - recent celery/kombu updates. 

I think if this is confirmed for 2.4.0, we will have in general more capacity to look more closely to such issues and see if this can be reproduced and fixed.

**Comment by V0lantis** [comment]
We will try this @potiuk, thank you !

**Comment by potiuk** [comment]
It's been released BTW :)

**Comment by hterik** [comment]
We got similar issue (Airflow 2.3.3), it happened right after worker suffered connection loss to Redis who later did not consume any tasks for 3 days, despite showing up as healthy in Flower. 
One thing i suspect is that reconnect attempt got stuck in an infinite timeout. Python sockets have default no timeout, same default applies in Celery and redis-py. (This would not explain why flower showed node as healthy though)

Discussed in https://github.com/redis/redis-py/issues/722 and https://github.com/redis/redis-py/issues/2243

In airflow, the celery_broker_transport_options option can be used to set the options discussed there.

**Comment by nicnguyen3103** [comment]
Hi we have a similar issue on Airflow 2.4.0. The issue happens after I set the pool too high, which leads to locking issue on Databases (the exact error was "Lock wait timeout exceeded; try restarting transaction"). The task was unable to mark failed for a while and after they did, the pool was kinda corrupted, which leads to the task being stuck in the queue. 

I do not have any direction to troubleshoot this issue at the moment, but if someone could give me pointer where to check, I am glad to help troubleshooting the issue 
<img width="749" alt="image" src="https://user-images.githubusercontent.com/46189714/193069502-38730235-cd28-47ae-b9a2-922058163fa8.png">

**Comment by Imcodingcc** [comment]
https://airflow.apache.org/docs/apache-airflow/stable/howto/dynamic-dag-generation.html#optimizing-dag-parsing-delays-during-execution

**Comment by Imcodingcc** [comment]
https://airflow.apache.org/docs/apache-airflow/stable/howto/dynamic-dag-generation.html#optimizing-dag-parsing-delays-during-execution 
in my case, I use kubernetes executor and dynamic DAG, I found it is stuck in DAG import in each task, I followed the example suggested by the official documentation, skipped  other DAG scans, and finally solved the problem

**Comment by fbmatilla** [comment]
Hi, we are facing this issue in production too. Hoping to have the fix too. Thanks.

**Comment by Walkerazos** [comment]
hi, we encountered this problem in production very fresh.  I hope you find a solution to the problem.  It's giving us a lot of trouble.  Thank you.

**Comment by patricker** [comment]
I don't think we've seen this since upgrading to 2.4.1.

**Comment by potiuk** [comment]
> hi, we encountered this problem in production very fresh. I hope you find a solution to the problem. It's giving us a lot of trouble. Thank you.

You have not explained which version you have. Did you upgrade to latest 2.4 ? version @Walkerazos ? This is anyhow the only way how fix can be applied (by upgrading to newer version) - and if you upgrade to 2.4 and you stop seing the problem, please report it here - we alredy got a few reports that the error has been solved since (there were multiple hundreds of errors fixed since and some of them related, so this is absolutely easiest and best way to verify if that problem has been fixed - by having several users like you who have this problem and it is gone after the upgrade.

Can you help us with that @Walkerazos ?

**Comment by rodrigoassumpcao** [comment]
Hi guys. I have the same issue.
I'm using KubernetesExecutor (on Azure), airflow version 2.4.2 (helm chart 1.7.0), kubernetes 1.24.6

Anyone knows a workaround for KubernetesExecutor?

This issue is causing a lot of troubles

**Comment by potiuk** [comment]
> Hi guys. I have the same issue. I'm using KubernetesExecutor (on Azure), airflow version 2.4.2 (helm chart 1.7.0), kubernetes 1.24.6
> 
> Anyone knows a workaround for KubernetesExecutor?
> 
> This issue is causing a lot of troubles

Then open a new issue and provice as much of an input as possible, showing your logs and circumstances (which might be very differnet from this case despite similarlities). This is the only way you can help with diagnosing your issue (and it is all in your hands to provide as much useful information as possible @rodrigoassumpcao )

**Comment by fbmatilla** [comment]
:( Looks like this is surfing from one milestone to the next one....

**Comment by potiuk** [comment]
And more oten than not, the reasons for observed behaaviour MIGHT be different in different versions. The only way to know if your issue (Which might seem siimilar) is actually the same as other's issue and to help with investigations of it is you  spending time on documenting it and providing evidences. You and your company paid exactly 0 USD for the software and helping community to investigate your problem is the LEAST you can do to contribute back for the absolutely free software you got, that many people developed in their free time. 

Not understanding this and complaining that you need to spend some time on documenting your issue (and putting thumbs down on comments of those people whe actually spend a ton of their free time on making the software avaialble for you for free shows lack of respect and empathy on your side @fbmatilla

**Comment by fbmatilla** [comment]
I was just reading all the thread an seeing this particular issue is not being very well managed. That's all. We can live without a fix, for sure. Thank you for all your time.

**Comment by potiuk** [comment]
> I was just reading all the thread an seeing this particular issue is not being very well managed. That's all. We can live without a fix, for sure. Thank you for all your time.

No. you were also downvoting comments without understanding the context of the project - which is (again let me repeat it) lack of respect and empathy.

**Comment by fbmatilla** [comment]
Yes, that's because I saw comments confirming this was solved but at the same time I could see the issue open and going from one milestone to the next one. I remove my downvoting if that's a problem....

**Comment by potiuk** [comment]
Apologies for doing that would be more appropriate - because you were downvoting the comments that explained why this was happening in response to your concern.

**Comment by fbmatilla** [comment]
Sure, apologies to all those working for free in this great and wonderful project!
And just remarking what I didn't like on this particular thread.
Thanks

**Comment by dharmendrakariya** [comment]
> Hi guys. I have the same issue. I'm using KubernetesExecutor (on Azure), airflow version 2.4.2 (helm chart 1.7.0), kubernetes 1.24.6
> 
> Anyone knows a workaround for KubernetesExecutor?
> 
> This issue is causing a lot of troubles

did you figure out bro? I am stuck with this!

**Comment by chenzuoli** [comment]
@Chris7 

> Simplify logic to resolve tasks stuck in queued despite stalled_task_timeout #30375

hello, I was wondering where do you place the code to run, add to a python script in the loop?

**Comment by chenzuoli** [comment]
> This is not ideal, but for those who want to kick all their queued tasks to running, here's a snippet i've been using:
> 
> ```
> from airflow import models, settings
> from airflow.executors.executor_loader import ExecutorLoader
> 
> session = settings.Session()
> tis = session.query(models.TaskInstance).filter(models.TaskInstance.state=='queued')
> dagbag = models.DagBag()
> for ti in tis:
>   dag = dagbag.get_dag(ti.dag_id)
>   task = dag.get_task(ti.task_id)
>   ti.refresh_from_task(task)
>   executor = ExecutorLoader.get_default_executor()
>   executor.job_id = "manual"
>   executor.start()
>   executor.queue_task_instance(ti, ignore_all_deps=False, ignore_task_deps=False, ignore_ti_state=False)
>   executor.heartbeat()
> ```

hello @Chris7 , I was wondering where do you place the code to run, add to a python script in the loop?

**Comment by funes79** [comment]
@chenzuoli @Chris7 this code works but in my case the job_id needs to be an integer, so something like this worked for me: 
```
...
  executor = ExecutorLoader.get_default_executor()
  executor.job_id = 1234567890
  executor.start()
...
```

Issue on 2.6.3 with posgre db backend, hundreds of tasks running in the same time in Kubernetes. No Celery. Usually it happens during those rush periods.

**Comment by gurmeetsaran** [comment]
Thanks @chenzuoli for providing the script. We use Airflow 2.2.2 and initially use the cronjob to run the provided script automatically to run task which were stuck. The script works great for non sensor related task but for sensor(specially reschedule mode), it becomes responsibility of script to manage the task for every single run. In a way the script becomes mini scheduler. This was becoming unmanageable so we decided to port the fix to  airflow 2.2.2 

The patch is available https://github.com/gurmeetsaran/airflow/pull/1.patch for airflow 2.2.2 (only celery executor). We are not seeing in task stuck issue after applying this patch.

Sharing it here if it helps anyone.
